### PR TITLE
Move `StorageController` persist writers to a Tokio task

### PR DESCRIPTION
**Pull request by frankmcsherry**
This PR moves the `WriteHandle`s from the storage controller implementation to a tokio task, which allows folks who call `append` to either block or not block on the returned future (a `oneshot::Receiver`). This allows certain unopinionated write frontier advancement to happen off of the main thread, while still allowing group commit to block. However, it also allows group commit to *not* block, if we would like to tweak that implementation (i.e. allow the future to prompt a future completion for the group commit, rather than blocking the main thread while it happens).

cc: @jkosh44 on some of these ideas.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/persist/src/file.rs [hunk 1]
```diff
@@ -125,6 +125,7 @@ impl Blob for FileBlob {
     }
 
     async fn set(&self, key: &str, value: Bytes, atomic: Atomicity) -> Result<(), ExternalError> {
+        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
```
Comment (jkosh44) [inline src/persist/src/file.rs:128]: Did you mean to leave this in?

File: src/adapter/src/coord.rs [hunk 2]
```diff
@@ -1269,7 +1274,8 @@ impl<S: Append + 'static> Coordinator<S> {
             })
             .collect::<Vec<_>>();
         let num_updates = appends.len();
-        self.controller.storage_mut().append(appends).await.unwrap();
+        // Note: Do not await the result; let it drop (as we don't need to block on completion)
+        self.controller.storage_mut().append(appends);
```
Comment (jkosh44) [inline src/adapter/src/coord.rs:1278]: If we're now doing these asynchronously now, should we still be queuing the updates and sending them in batches? It seems fine to fire them all at once and forget about them.

**Review by jkosh44 — COMMENTED** [review]
I didn't realize this was still in draft until after I finished looking at it. I just saw the notification and took a look.

One thing to keep in mind is that the asynchronous table advancements might fail if a non-empty write ends up arriving at persist first. That's totally fine and correct, we should just make sure things don't blow up.

File: src/adapter/src/coord.rs (hunk 2 above)
Comment (frankmcsherry) [reply src/adapter/src/coord.rs:1278]: If we do that I should edit it a bit to consolidate the writes in the tokio task. But I wanted to randomize things in the coordinator as little as possible.

File: src/persist/src/file.rs (hunk 1 above)
Comment (frankmcsherry) [reply src/persist/src/file.rs:128]: Hah, no.

**Comment by frankmcsherry** [comment]
> One thing to keep in mind is that the asynchronous table advancements might fail if a non-empty write ends up arriving at persist first. That's totally fine and correct, we should just make sure things don't blow up.

Yeah, actually the task should have a process perfect view of what has been shipped, so it could maintain some "intended frontiers" and do some pre-filtering that `persist` would otherwise do for it; that is a good idea. It can also consolidate them, if there are a few outstanding advancements to do (we should be able to do them much more frequently now, too).

File: src/storage/src/controller.rs [hunk 3]
```diff
@@ -843,15 +777,17 @@ where
...
+        let msg = tokio::select! {
+            // Order matters here. We want to process internal commands
+            // before processing external commands.
+            biased;
+
+            Some(m) = self.internal_response_queue.recv() => m,
+            Some((_id, m)) = clients.next() => m,
+        };
+
+        self.state.stashed_response = Some(msg);
```
Comment (frankmcsherry) [inline src/storage/src/controller.rs:790]: @benesch I just made this stuff up and would love your opinion. The intent is that there is now always an `internal_response_queue` to await, so the "no clients" issue shouldn't exist. I'm not sure I'm respecting cancelation safety and such, but it looks like the `StreamMap` probably wasn't either so hoping it's ok.

Comment (benesch) [reply src/storage/src/controller.rs:790]: This bit LGTM! We're now mandating that the clients are cancel safe so this is all good.

**Review by jkosh44 — APPROVED** [review]
Coord stuff LGTM. I'm not really familiar enough with the storage controller to sign off on that.

**Comment by frankmcsherry** [comment]
cc: @aljoscha and @petrosagg to take a peek once they return!

**Comment by mjibson** [comment]
I would like to know how this interacts with the advance local inputs "try to make stuff take 50ms" code. If these now don't await, then I assume that code will get to "always send all table advancements at once". Let's assume the total time it takes to do all advancements exceeds 1s (at which point coord will send a new batch). What exactly happens on the storage side? Do those commands get backed up and cause other storage commands to block? One of the very nice properties of today's coord advancement is if this happens, coord will skip over the old updates and move right to the newer ones (a cheap form of consolidation). Does storage have a similar property? I worry this is moving one performance problem into another domain, and the optimizations we made will now make things worse.

**Comment by frankmcsherry** [comment]
> I would like to know how this interacts with the advance local inputs "try to make stuff take 50ms" code.

Great question! I don't know how that code works. The way the Tokio task works, it peels off all `Append` commands, consolidates all of them, and attempts to apply them all in one `compare_and_append` call. This means that if there are 1,000 "table advance" commands outstanding, they turn in to one `compare_and_append` for each table. Clearly, we might prefer that they not be 1,000 in the first place, but the cost of having random extras shouldn't be too high.

**Comment by mjibson** [comment]
Ok, that sounds like it'll do an ok thing (consolidation, which I assume for this message means "if you see multiple advance messages for the same table, choose the one with the largest timestamp") if advancing all tables takes longer than 1 second. What happens to other storage commands during the time when all tables are being advanced? Are they blocked.

**Comment by frankmcsherry** [comment]
> if advancing all tables takes longer than 1 second. What happens to other storage commands during the time when all tables are being advanced? Are they blocked?

Ah, the storage controller should be immediately available after an `append` call. What will happen if it takes a while to write to `persist` is that the queue of commands may back up, which means the next time we drain it there will be that many more commands. If we saturate the `environmentd` to `persist` bandwidth, the queue will back further and further up, and eventually we will explode (there is no backpressure built in). If we want to, we could fix that by having some time-out, and just marking any write that ends up more stale than that time-out as failed and responding as such.

**Comment by mjibson** [comment]
I'm wondering if things like "insert one row into a table" could now take a very long time because table advancement is no longer being split up. The goal with the current code is to make sql queries seem fast, by making tiny batches of work. If an insert can now block for however long it takes to advance all tables, then we are back to where we were.

Here's a thing I think we could do: teach coord's table advancement to await the append in a new tokio task, and then make its batching size decisions based on that. This would preserve the current perf intentions while also moving work off of the coord thread.

**Comment by frankmcsherry** [comment]
With the last addition, this should fix #13454 (it does so locally). The issue was that the reporting of the changes to `upper` frontiers was not based on `T::minumum()` but whatever `WriteHandle::upper()` started at.

File: src/adapter/src/coord.rs [hunk 4]
```diff
@@ -1269,7 +1275,12 @@ impl<S: Append + 'static> Coordinator<S> {
             })
             .collect::<Vec<_>>();
         let num_updates = appends.len();
-        self.controller.storage_mut().append(appends).await.unwrap();
+        // Note: Do not await the result; let it drop (as we don't need to block on completion)
+        // The error that could be return
```
Comment (aljoscha) [inline src/adapter/src/coord.rs:1279]: Something seems to be missing here?

**Review by aljoscha — COMMENTED** [review]
This looks very good! :+1:

**Pull request merged**